### PR TITLE
CSRMM Library Node

### PR DESCRIPTION
**Pull request by lukastruemper**

### CODE REVIEW

**Comment by lukastruemper** [comment]
@alexnick83 cuSPARSE breaks when using the "+ beta * C" part and pure lacks the "transpose B" part. Other than that, it should be good

File: dace/libraries/sparse/nodes/csrmm.py [hunk 1]
```diff
@@ -0,0 +1,688 @@
...
+                beta = f'{dtype.ctype}({node.beta.real}, {node.beta.imag})'
+            else:
+                beta = f'{dtype.ctype}({node.beta})'
+
+            # Set pointer mode to host
+            call_prefix += f'''cusparseSetPointerMode(__dace_cusparse_handle, CUSPARSE_POINTER_MODE_HOST);
+            {dtype.ctype} alpha = {alpha};
+            {dtype.ctype} beta = {beta};
+            '''
+            call_suffix += '''cusparseSetPointerMode(__dace_cusparse_handle, CUSPARSE_POINTER_MODE_DEVICE);'''
```
Comment (tbennun) [inline dace/libraries/sparse/nodes/csrmm.py:399]: why do both?

File: dace/libraries/sparse/nodes/csrmm.py [hunk 2]
```diff
@@ -0,0 +1,688 @@
...
+            dace::sparse::CheckCusparseError( cusparseCreateDnMat(&matC, {arows}, {bcols}, {ldc}, {arr_prefix}_c,
+                                                {compute}, {layout}) );
+            // allocate an external buffer if needed
+            dace::sparse::CheckCusparseError( cusparseSpMM_bufferSize(
+                                            {handle},
+                                            {opA},
+                                            {opB},
+                                            {alpha}, matA, matB, {beta}, matC, {compute},
+                                            CUSPARSE_SPMM_ALG_DEFAULT, &bufferSize) );
+            cudaMalloc(&dBuffer, bufferSize);
```
Comment (tbennun) [inline dace/libraries/sparse/nodes/csrmm.py:469]: I really hate this. I know it was in the original implementation but it's an uncalled-for malloc/free

File: dace/libraries/sparse/nodes/csrmm.py [hunk 3]
```diff
@@ -0,0 +1,688 @@
...
+            cusparseDnMatDescr_t matB, matC;
+            void*                dBuffer    = NULL;
+            size_t               bufferSize = 0;
+            // Create sparse matrix A in CSR format
+            dace::sparse::CheckCusparseError( cusparseCreateCsr(&matA, {arows}, {acols}, {annz},
+                                                {arr_prefix}_a_rows, {arr_prefix}_a_cols, {arr_prefix}_a_vals,
+                                                CUSPARSE_INDEX_32I, CUSPARSE_INDEX_32I,
+                                                CUSPARSE_INDEX_BASE_ZERO, {compute}) );
+            // Create dense matrix B
+            dace::sparse::CheckCusparseError( cusparseCreateDnMat(&matB, {acols}, {bcols}, {ldb}, {arr_prefix}_b,
```
Comment (tbennun) [inline dace/libraries/sparse/nodes/csrmm.py:457]: is there no way to create those descriptors in the init code or something?

File: dace/libraries/sparse/nodes/csrmm.py [hunk 4]
```diff
@@ -0,0 +1,688 @@
...
+        #         {arr_prefix}{opt['y']},
+        #         {dtype_to_cudadatatype(opt['ydtype'])},
+        #         {opt['ldb']},
+        #         {beta},
+        #         {arr_prefix}_c,
+        #         {dtype_to_cudadatatype(opt['cdtype'])},
+        #         {opt['ldc']},
+        #         {acctype},
+        #         {algorithm});
+        #     '''
```
Comment (tbennun) [inline dace/libraries/sparse/nodes/csrmm.py:525]: can remove

File: dace/libraries/sparse/nodes/csrmm.py [hunk 5]
```diff
@@ -0,0 +1,688 @@
...
+                                desc="A scalar which will be multiplied with A @ B before adding C")
+    beta = properties.Property(allow_none=False,
+                               default=0,
+                               desc="A scalar which will be multiplied with C before adding C")
+    cin = properties.Property(dtype=bool, default=True, desc="Whether to have a _cin connector when beta != 0")
+    algorithm = properties.Property(dtype=str,
+                                    allow_none=True,
+                                    default=None,
+                                    desc="If applicable, chooses the vendor-provided implementation "
+                                    "(algorithm) for the multiplication")
```
Comment (tbennun) [inline dace/libraries/sparse/nodes/csrmm.py:618]: not used. Is it possible to use it?

File: dace/libraries/sparse/nodes/csrmm.py [hunk 6]
```diff
@@ -0,0 +1,688 @@
...
+    compute_type = properties.Property(default=None,
+                                       dtype=str,
+                                       allow_none=True,
+                                       desc="If applicable, overrides computation type (CUBLAS-specific, see "
+                                       "``cublasComputeType_t``)")
```
Comment (tbennun) [inline dace/libraries/sparse/nodes/csrmm.py:623]: same. if not possible to use, please remove

**Review by tbennun — COMMENTED** [review]
questions mostly

**Comment by alexnick83** [comment]
> @alexnick83 cuSPARSE breaks when using the "+ beta * C" part and pure lacks the "transpose B" part. Other than that, it should be good

Is it a cuSPARSE issue or is our code broken?

File: dace/libraries/sparse/nodes/csrmm.py (hunk 2 above)
Comment (alexnick83) [reply dace/libraries/sparse/nodes/csrmm.py:469]: How will you avoid this? I guess we need to use the memory pool, but I haven't looked at the API for that.

**Comment by lukastruemper** [comment]
> > @alexnick83 cuSPARSE breaks when using the "+ beta * C" part and pure lacks the "transpose B" part. Other than that, it should be good
> 
> Is it a cuSPARSE issue or is our code broken?

No, it's the "in" connector that is not handled for cuSPARSE. When we create the device-arrays for GPU in the nested SDFG, "in" is not created. I tried to add this but then in the cusparse code, we need to change the buffer to cin for Mat C. If I do so, the result is written to cin and I don't know what would be the smartest way to copy this back to "cout"/"c"

**Review by tbennun — APPROVED** [review]

File: dace/libraries/sparse/nodes/csrmm.py [hunk 7]
```diff
@@ -524,13 +533,6 @@ def expansion(node, state, sdfg):
             nstate.add_edge(tasklet, '_conn_c', gc, None, dace.Memlet.from_array('_c_gpu', cdesc))
             nstate.add_nedge(gc, c, dace.Memlet.from_array('_c', cdesc))
 
-            if node.beta != 0.0:
```
Comment (tbennun) [inline dace/libraries/sparse/nodes/csrmm.py]: why was this removed?

Comment (lukastruemper) [reply dace/libraries/sparse/nodes/csrmm.py]: > why was this removed?

Because I added a pure-Expansion + GPUTransformSDFG for the case of beta != 0 before the outer if statement

I can split this into two commits?

File: dace/libraries/sparse/nodes/csrmm.py [hunk 8]
```diff
@@ -463,9 +483,16 @@ def expansion(node, state, sdfg):
 
         # If buffers are not on the GPU, copy them
         if needs_copy:
+            if node.beta != 0.0:
```
Comment (lukastruemper) [inline dace/libraries/sparse/nodes/csrmm.py:486]: @tbennun

Comment (tbennun) [reply dace/libraries/sparse/nodes/csrmm.py:486]: somehow the diff missed it. thanks!
